nifi-framework-api/src/main/java/org/apache/nifi/nar/NarPersistenceProvider.java

```diff
@@ -0,0 +1,109 @@
+     */
+    void initialize(NarPersistenceProviderInitializationContext initializationContext) throws IOException;
+
+    /**
+     * Creates a temporary file containing the contents of the given input stream.
+     *
+     * @param inputStream the input stream
+     * @return the file containing the contents of the input stream
+     *
+     * @throws IOException if an I/O error occurs writing the stream to the temp file
+     */
+    File createTempFile(InputStream inputStream) throws IOException;
+
+    /**
+     * Saves the temporary NAR file to its final destination.
+     *
+     * @param persistenceContext the context information
+     * @param tempNarFile the temporary file created from calling createTempFile
+     * @return the information about the persisted NAR, including the saved file
+     *
+     * @throws IOException if an I/O error occurs saving the temp NAR file
+     */
+    NarPersistenceInfo saveNar(NarPersistenceContext persistenceContext, File tempNarFile) throws IOException;
+
+    /**
+     * Deletes the contents of the NAR with the given coordinate.
+     *
+     * @param narCoordinate the coordinate of the NAR to delete
+     *
+     * @throws FileNotFoundException if a NAR with the given coordinate does not exist
+     * @throws IOException if an I/O error occurred deleting the NAR
+     */
+    void deleteNar(BundleCoordinate narCoordinate) throws IOException;
+
+    /**
+     * Obtains an input stream to read the contents of the NAR with the given coordinate.
+     *
+     * @param narCoordinate the coordinate of the NAR
+     * @return the input stream for the NAR
+     *
+     * @throws FileNotFoundException if a NAR with the given coordinate does not exist
+     */
+    InputStream readNar(BundleCoordinate narCoordinate) throws FileNotFoundException;
+
+    /**
+     * Retrieves the persistence info for the NAR with the given NAR coordinate.
+     *
+     * @param narCoordinate the coordinate of the NAR
+     * @return the persistence info
+     * @throws IOException if an I/O error occurs reading the persistence info
+     */
+    NarPersistenceInfo getNarInfo(BundleCoordinate narCoordinate) throws IOException;
+
+    /**
+     * Indicates if a NAR with the given coordinate exists in this persistence provider.
+     *
+     * @param narCoordinate the coordinate of the NAR
+     * @return true if a NAR with the coordinate exists, false otherwise
+     */
+    boolean exists(BundleCoordinate narCoordinate);
+
+    /**
+     * @return the info for all NARs persisted by this persistence provider
+     */
+    Set<NarPersistenceInfo> getAllNarInfo() throws IOException;
+
+    /**
+     * Shuts down the provider and cleans up any resources held by it. Once this method has returned, the
+     * provider may be initialized once again via the {@link #initialize(NarPersistenceProviderInitializationContext)} method.
+     */
+    void shutdown();
```

Comment:
What do you think about calling this `close()` and extending the `Closeable` interface. The only downside is the potential for throwing an `IOException`, but that may be useful in an implementation.

Reply:
It is never really intended to be closed except when shutting down the application and seemed more consistent with other framework extensions like `StateProvider` and repositories which have methods like `shutdown()` or `cleanup()`. I lean towards leaving this as is.

Reply:
Thanks, that makes sense, the current approach looks good.